design/spark-co-exist-with-object-storages.md

```diff
@@ -0,0 +1,141 @@
+
+### URI translation
+
+`RouterFileSystem` has access to a configurable mapping that maps bucket names into lakeFS repositories and prefixes, e.g.
+```yaml
+MapperConfig = {
+    {
+        "source_bucket_pattern": "bucket-a",
+        "mapped_bucket_name": "example-repo",
+        "mapped_prefix": "dev/"
+    }
+}
+```
+
+With the mapping above, the URI `s3://bucket-a/foo.parquet` will be translated into `s3://example-repo/dev/foo.parquet`. 
+
+### Invoke file system operations
+
+After translating URIs to their final form, `RouterFileSystem` will use the translated path and its relevant file system to 
+perform file system operations against the relevant object store. 
+
+Considering the example above, the relevant object store and file system are S3 and `S3AFileSystem` compatibly. `RouterFileSystem` will use `S3AFileSystem` to perform file system operations against S3. e.g. to open `s3://example-repo/dev/foo.parquet`
+
+#### Getting the relevant File System 
+
+Given that `RouterFileSystem` is configured as the file system for URIs with a certain scheme (e.g. `scheme=s3`),
+we need to provide a way to get the relevant file system that can perform actual file system operations against the 
+relevant object store, e.g. `S3AFileSystem`. 
+To do that, we would follow a two-step process: 
+1. add the following Spark configuration:
+```shell
+# fs.underlying.impl=relevant filesystem
+fs.underlying.impl=S3AFileSystem
+``` 
+The configuration above tells Spark that `S3AFileSystem` is the relevant file system for URIs with `scheme=underlying`. 
+
+2. Use a URI with `scheme=underlying` to fetch the relevant file system. 
+```java
+String origPath = "s3://example-repo/dev/foo.parquet";
+String fsGetterPath = "underlying://example-repo/dev/foo.parquet";
+FileSystem underlyingFS = fsGetterPath.getFileSystem(); // S3AFileSystem according to Spark configurations
+// Can later do underlyingFS.open(origPath) 
+```
+
+### Integrating with lakeFS
+
+`RouterFileSystem` does not change the exiting [integration methods](../docs/integrations/spark.md#two-tiered-spark-support) 
+lakeFS and Spark have. 
+While accessing lakeFS [using the S3 gateway](../docs/integrations/spark.md#access-lakefs-using-the-s3a-gateway) we need 
+to use [per-bucket configurations](../docs/integrations/spark.md#per-bucket-configuration) so that the relevant underlying 
+file system can direct traffic to the lakeFS server for mapped buckets. 
+
+#### Example: Read an object managed by lakeFS
+
+![RouterFS with lakeFS URI](diagrams/routerFS_lakefs_path.png)
+
+### How RouterFileSystem works with URIs that are not mapped to lakeFS objects   
+
+`RouterFileSystem` will check the mapping configurations, and won't change paths that are not mapped to lakeFS objects. 
+e.g. given the URI `s3://bucket-b/bar.parquet` and the following mapping configurations:
+```yaml
+MapperConfig = {
```

Comment:
1. consider using properties - trying to act like the ecosystem
2. can we simplify it to source prefix and target prefix? the suggestion forces a full bucket as the source.

Reply:
1. done
2. let me know what you think now